### PR TITLE
sys: add mutex lock with timeout function

### PR DESCRIPTION
**Pull request by lebrush**
This implements mutex locking with a timeout, which is required for the semaphore implementation without IPC (#6155). 

It's been split into a separate PR as [suggested](https://github.com/RIOT-OS/RIOT/pull/6155#discussion_r89370960) by @miri64

Depends on: #6211
Required by: #6155

### CODE REVIEW

File: core/include/thread.h [hunk 1]
```diff
@@ -434,6 +434,19 @@ char *thread_stack_init(thread_task_func_t task_func, void *arg, void *stack_sta
  */
 void thread_add_to_list(list_node_t *list, thread_t *thread);
 
+/**
+ * @brief Remove thread from list (internal)
+ *
+ * This will remove @p thread from @p list. It reuses the thread's rq_entry
+ * field. Used internally timeout mutex implementation.
```
Comment (miri64) [inline core/include/thread.h]: Can you directly reference the function here?

```
Used internally by @ref xtimer_mutex_lock_timeout() "timeout mutex implementation".
```

File: core/include/thread.h [hunk 2]
```diff
@@ -434,6 +434,19 @@ char *thread_stack_init(thread_task_func_t task_func, void *arg, void *stack_sta
  */
 void thread_add_to_list(list_node_t *list, thread_t *thread);
 
+/**
+ * @brief Remove thread from list (internal)
```
Comment (OlegHahm) [inline core/include/thread.h]: Maybe `@internal`?

**Comment by OlegHahm** [comment]
@BytesGalore, didn't we have a discussion about this feature some years before?

File: core/include/thread.h (hunk 2 above)
Comment (lebrush) [inline core/include/thread.h]: [`thread_add_to_list`](https://github.com/RIOT-OS/RIOT/pull/6158/files/94c15e16c70fec51df16af3f3dcd9256e22f2bb5#diff-9dcf1c99b916850378a11faa9f013f32R423) is not @internal. Should I update both?

**Comment by lebrush** [comment]
What was the outcome of the discussion @OlegHahm ? :-)

**Comment by BytesGalore** [comment]
@OlegHahm, @lebrush I really can't remember, but I guess it was related to the requirement of have condition variables. Since solely using mutex could result in a race and in some cases in a deadlock.

**Comment by lebrush** [comment]
I see... I don't have such a deep overview of the `sched` and `thread` modules as @kaspar030 and @BytesGalore, can you guys think of any race/deadlock condition with the proposed implementation?

Also, I've to see if I've to rebase to accommodate the latest changes in the xtimer.

**Comment by kaspar030** [comment]
> @kaspar030 and @BytesGalore, can you guys think of any race/deadlock condition with the proposed implementation?

No, @BytesGalore, can you elaborate?

**Comment by lebrush** [comment]
What is holding this back? I'd be nice to have this and #6155 merged for the next release

File: core/thread.c [hunk 3]
```diff
@@ -124,6 +124,19 @@ void thread_add_to_list(list_node_t *list, thread_t *thread)
     list->next = new_node;
 }
 
+void thread_remove_from_list(list_node_t *list, thread_t *thread)
```
Comment (kaspar030) [inline core/thread.c]: Why not use clist_remove()?

File: sys/xtimer/xtimer.c [hunk 4]
```diff
@@ -217,7 +223,36 @@ int xtimer_msg_receive_timeout(msg_t *msg, uint32_t us)
...
     xtimer_t t;
+
     _setup_timer_msg(&tmsg, &t);
     xtimer_set_msg(&t, us, &tmsg, sched_active_pid);
     return _msg_wait(msg, &tmsg, &t);
 }
+
+static void _mutex_timeout(void *arg)
+{
+    unsigned irqstate = irq_disable();
```
Comment (kaspar030) [inline sys/xtimer/xtimer.c]: This is run in ISR context, so no need for disabling IRQs.

File: sys/xtimer/xtimer.c [hunk 5]
```diff
@@ -217,7 +223,36 @@ int xtimer_msg_receive_timeout(msg_t *msg, uint32_t us)
 {
     msg_t tmsg;
     xtimer_t t;
+
```
Comment (kaspar030) [inline sys/xtimer/xtimer.c]: unrelated change

File: sys/xtimer/xtimer.c [hunk 6]
```diff
@@ -32,6 +32,12 @@
 #define ENABLE_DEBUG 0
 #include "debug.h"
 
+typedef struct {
+    mutex_t *mutex;
+    thread_t *thread;
+    int timeout;
```
Comment (kaspar030) [inline sys/xtimer/xtimer.c]: what is this used for?

File: core/thread.c (hunk 3 above)
Comment (lebrush) [inline core/thread.c]: I didn't consider it because: the list is not a circular list and because when added is using a `thread_add_to_list` which is in the thread module (I assume that this is done this way to lower inter-dependencies).

File: sys/xtimer/xtimer.c (hunk 6 above)
Comment (lebrush) [inline sys/xtimer/xtimer.c]: When the mutex times out, this flag is set in order to indicate whether it timed out or not (see L257).

File: sys/xtimer/xtimer.c (hunk 4 above)
Comment (lebrush) [inline sys/xtimer/xtimer.c]: You do need, otherwise you might run in a race condition in case an interrupt occurs while serving this interrupt. Imagine: an that the other interrupt occurs while removing the mutex from the queue and that interrupt does the same operation (i.e. by calling `mutex_unlock`).

**Comment by lebrush** [comment]
@kaspar030 rebased onto latest master and squashed changes

File: core/thread.c (hunk 3 above)
Comment (miri64) [inline core/thread.c]: Alternatively: why not add a `list_remove()` implementation? Seems a lot cleaner.

Comment (kaspar030) [inline core/thread.c]: thread_add_to_list() is taking the thread's priority into account and adds the thread so afterwards the list is sorted by priority.

+1 for adding "list_remove()" to list.h! and using that!

Comment (kaspar030) [inline core/thread.c]: > adding "list_remove()" to list.h

It's probably best to do that in a seperate PR.

File: sys/xtimer/xtimer.c (hunk 4 above)
Comment (kaspar030) [inline sys/xtimer/xtimer.c]: @lebrush RIOT currently doesn't allow nested interrupts, so while one ISR is running, no other will interrupt.

File: core/thread.c (hunk 3 above)
Comment (lebrush) [inline core/thread.c]: Agree! Will do :-) now we'll have a PR which depends in this PR which depends in another PR, yeeey :tada:

File: sys/xtimer/xtimer.c (hunk 4 above)
Comment (lebrush) [inline sys/xtimer/xtimer.c]: Again what learnt ;-) Will fix

File: core/thread.c (hunk 3 above)
Comment (lebrush) [inline core/thread.c]: Now this will depend on #6211

Comment (miri64) [inline core/thread.c]: > Agree! Will do :-) now we'll have a PR which depends in this PR which depends in another PR, yeeey :tada:

Welcome to RIOT ;-P

**Comment by miri64** [comment]
@kaspar030 ping?

**Comment by miri64** [comment]
#6211 got merged

**Comment by lebrush** [comment]
Rebased. @kaspar030 any other change requested?

**Comment by kaspar030** [comment]
ACK.

**Comment by lebrush** [comment]
Can anyone trigger Murdock again, please?

```
Running './dist/tools/pr_check/pr_check.sh master' x
Command output:

	Pull request is waiting for another pull request according to its labels set on GitHub
```
